### ISSUE TITLE
hpilo_ca sign needs a DNS alias.

### ISSUE THREAD

**Issue by duncanwebb** [seveas/python-hpilo#80]
Your python-hpilo is just great and has saved me many hours.

I've added a few code changes so the hpilo_ca sign works against a iLO 4, I'm not sure how the generate a patch using git against the released version.

One thing that would save even more time is the generation of the certificates that work with the onboard administrator. They open a new window using the IP address of the iLO and not the host name. This means that the certificate needs a DNS alias for the IP addess as well as the unqualified host name.

When signing the certificate the lines:
nsSslServerName = $ENV::NSSSLSERVERNAME
subjectAltName = $ENV::SUBJECTALTNAMERAW
allow environment variables to be used by openssl. 

SUBJECTALTNAMEDNS=DNS:10.30.24.193,DNS:chglbvaruat121-ilo

Sign it with using a section in the openssl.cnf file called server_ca:
/usr/bin/openssl ca -batch -passin env:SSLPASS -notext -config openssl.cnf -name server_ca -in "10.30.24.193.pem" -days 1826 -preserveDN

Would it be possible to add this functionality? In the meantime I'll carry on trying the figure out what to add and where into the code.

**Comment by seveas** [comment]
On ma, 2015-05-18 at 04:47 -0700, duncanwebb wrote:
> Your python-hpilo is just great and has saved me many hours.
> 
That's good to hear :)

> I've added a few code changes so the hpilo_ca sign works against a iLO
> 4, I'm not sure how the generate a patch using git against the
> released version.

Thanks for contributing back! I haven't actually used hpilo_ca sign yet
with an ilo4 so I'm curious to see what the problems are.

To generate a patch for uncommitted changes: git diff
To generate a patch for changes you already committed: git format-patch origin/master

> One thing that would save even more time is the generation of the
> certificates that work with the onboard administrator. They open a new
> window using the IP address of the iLO and not the host name. This
> means that the certificate needs a DNS alias for the IP addess as well
> as the unqualified host name.

That would be a nice idea. I don't generally visit iLO's this way, but
can see that it would help in other use-cases. It would require a new
certificate when the IP address changes though, but that's not a big
deal.

There's also the problem that the iLO generates the CSR, so we have to
mangle it to include the subjectAltName (and hope the iLO actually
supports this.

-- 
Dennis Kaarsemaker
http://www.kaarsemaker.net

**Comment by duncanwebb** [comment]
On 18/05/2015 15:40, Dennis Kaarsemaker wrote:
> On ma, 2015-05-18 at 04:47 -0700, duncanwebb wrote:
>
> > I've added a few code changes so the hpilo_ca sign works against a iLO
> > 4, I'm not sure how the generate a patch using git against the
> > released version.
>
> Thanks for contributing back! I haven't actually used hpilo_ca sign yet
> with an ilo4 so I'm curious to see what the problems are.
>
> To generate a patch for uncommitted changes: git diff
> To generate a patch for changes you already committed: git 
> format-patch origin/master

I was using the 2.9 release but the code hasn't changed much, so the 
patch may apply. One thing to note is that when requesting a CSR it can 
take the iLO about 12 mins to prepare one. I've only tried this with iLO 4.

>
> > One thing that would save even more time is the generation of the
> > certificates that work with the onboard administrator. They open a new
> > window using the IP address of the iLO and not the host name. This
> > means that the certificate needs a DNS alias for the IP addess as well
> > as the unqualified host name.
>
> That would be a nice idea. I don't generally visit iLO's this way, but
> can see that it would help in other use-cases. It would require a new
> certificate when the IP address changes though, but that's not a big
> deal.

With blade enclosures the on-board administrator is the easiest way to 
go as all 16 blades are accessible from one interface. From there we can 
click onto the individual iLOs and their remote consoles.

> There's also the problem that the iLO generates the CSR, so we have to
> mangle it to include the subjectAltName (and hope the iLO actually
> supports this.

This does work, I used hpilo_cli to generate the CSR, It seems that the 
iLO need the FQDN of the host to accept the certificate and any change 
to the network settings causes the certificate to reset to the default 
useless one. The CSR generated by the iLO includes a subjectAltName for 
the FQDN and the host name but not the IP address. So I have to add one 
when signing the request. I wasn't sure if this was possible with 
openssl but found the answer that it is possible (from TinyCA). I 
manually signed the certificate and imported it with the import 
certificate command. It worked and no more untrusted certificate.

BTW I think that HP have broken their RIBCL commands with iLO4 version 
2.10 as they now return an error saying "Error writing xml_header", I've 
a case with HP about this error message.

Kind regards
Duncan Webb

--- ../python-hpilo-2.13.1/examples/ca/hpilo_ca	2015-04-15 15:37:54.000000000 +0200
+++ hpilo_ca	2015-05-19 06:43:24.447786401 +0200
@@ -1,6 +1,6 @@
 #!/usr/bin/python
 #
-# (c) 2011-2015 Dennis Kaarsemaker <dennis@kaarsemaker.net>
+# (c) 2011-2013 Dennis Kaarsemaker <dennis@kaarsemaker.net>
 # see COPYING for license details
 
 try:
@@ -45,7 +45,7 @@
         p.exit(1)
 
     if not os.path.exists(os.path.expanduser(opts.config)):
-        p.error("COnfiguration file does not exist")
+        p.error("Configuration file '%s' does not exist" % opts.config)
 
     config = ConfigParser.ConfigParser()
     config.read(os.path.expanduser(opts.config))
@@ -93,13 +93,14 @@
     if opts.password:
         password = opts.password
     if not login or not password:
-        p.error("No login details provided")
+        p.error("No iLO login details provided")
 
+    pending_requests = []
     for hostname in args[1:]:
         csr_path = os.path.join(ca_path, 'certs', hostname + '.csr')
         crt_path = os.path.join(ca_path, 'certs', hostname + '.crt')
         if os.path.exists(crt_path):
-            print("Certificate already signed, skipping")
+            print("Certificate already signed for %s, skipping" % hostname)
             continue
 
         ilo = hpilo.Ilo(hostname, login, password, opts.timeout, opts.port)
@@ -131,27 +132,72 @@
                 ilo.cert_fqdn(use_fqdn=True)
                 cn = ilo.get_cert_subject_info()['csr_subject_common_name']
             if cn != hostname:
-                print("Hostname (%s) and CN (%s) do not match, fixing" % (hostname, cn))
-                ilo.mod_network_settings(dns_name=hostname[:hostname.find('.')])
-                print("Waiting a minute to let the iLO reset itself")
-                time.sleep(60)
+                print("Hostname (%s) and CN (%s) do not match, not fixing" % (hostname, cn))
+            #    print("Hostname (%s) and CN (%s) do not match, fixing" % (hostname, cn))
+            #    ilo.mod_network_settings(dns_name=hostname[:hostname.find('.')])
+            #    print("Waiting a minute to let the iLO reset itself")
+            #    time.sleep(60)
 
-        elif re.match(r'^iLO{3-9}$', fw_version['management_processor']):
+        elif re.match(r'^iLO[3-9]$', fw_version['management_processor']):
             cn = ilo.get_network_settings()['dns_name']
             if cn != hostname[:hostname.find('.')]:
-                print("Hostname (%s) and CN (%s) do not match, fixing" % (hostname, cn))
-                ilo.mod_network_settings(dns_name=hostname[:hostname.find('.')])
-                print("Waiting a minute to let the iLO reset itself")
-                time.sleep(60)
+                print("Hostname (%s) and CN (%s) do not match, not fixing" % (hostname, cn))
+            #    print("Hostname (%s) and CN (%s) do not match, fixing" % (hostname, cn))
+            #    ilo.mod_network_settings(dns_name=hostname[:hostname.find('.')])
+            #    print("Waiting a minute to let the iLO reset itself")
+            #    time.sleep(60)
 
         else:
-            print("hpilo_ca cannot manage certificates for %s, only iLO2 and newer are supported" % (fw_version['management_processor'], hostname))
+            print("hpilo_ca cannot manage certificates for %s %s, only iLO2 and newer are supported" % (fw_version['management_processor'], hostname))
+            continue
+
+        print("(2/5) Retrieving certificate signing request")
+        try:
+            # this is a bit poor as I have hardcoded the certificat information, not sure that is works
+            csr = ilo.certificate_signing_request()
+            #csr = ilo.certificate_signing_request(country="CH", state="ZH", locality="Zurich", organization="Leonteq", organizational_unit="Information Technology", common_name=hostname)
+        except hpilo.IloError as e:
+            print(str(e))
+            pending_requests.append(hostname)
+            continue
+        if not csr:
+            print("Received an empty CSR")
             continue
+        fd = open(csr_path, 'w')
+        fd.write(csr)
+        fd.close()
+
+        print("(3/5) Signing certificate")
+        openssl('ca', '-batch', '-in', csr_path, '-out', crt_path)
 
+        print("(4/5) Uploading certificate")
+        fd = open(crt_path)
+        cert = fd.read()
+        fd.close()
+        cert = cert[cert.find('-----BEGIN'):]
+        ilo.import_certificate(cert)
+
+        print("(5/5) Resetting iLO")
+        ilo.reset_rib()
+
+    print("*********************")
+    print("%s" % " ".join(pending_requests))
+    print("*********************")
+    if pending_requests:
+        print("Waiting 15 minutes to let the iLO generate the signing request")
+        for i in range(15):
+            time.sleep(60)
+
+    for hostname in pending_requests:
         print("(2/5) Retrieving certificate signing request")
-        csr = ilo.certificate_signing_request()
+        try:
+            csr = ilo.certificate_signing_request()
+            #csr = ilo.certificate_signing_request(country="CH", state="ZH", locality="Zurich", organization="Leonteq", organizational_unit="Information Technology", common_name=hostname)
+        except hpilo.IloError as e:
+            print(str(e))
+            continue
         if not csr:
-            print("Reveived an empty CSR")
+            print("Received an empty CSR")
             continue
         fd = open(csr_path, 'w')
         fd.write(csr)
@@ -170,6 +216,7 @@
         print("(5/5) Resetting iLO")
         ilo.reset_rib()
 
+
 class OpenSSL(object):
     def __init__(self, openssl_bin, openssl_cnf):
         self.openssl_bin = openssl_bin

**Comment by duncanwebb** [comment]
Here is anther patch based on the git, tested against iLO2 and iLO4. There is some code that is commented out to retry pending CSR requests. I'm not 100% sure which CSR fields are optional.
```
diff --git a/examples/ca/hpilo_ca b/examples/ca/hpilo_ca
index 71aba79..3950bad 100755
--- a/examples/ca/hpilo_ca
+++ b/examples/ca/hpilo_ca
@@ -45,7 +45,7 @@ def main():
         p.exit(1)
 
     if not os.path.exists(os.path.expanduser(opts.config)):
-        p.error("COnfiguration file does not exist")
+        p.error("Configuration file '%s' does not exist" % opts.config)
 
     config = ConfigParser.ConfigParser()
     config.read(os.path.expanduser(opts.config))
@@ -55,6 +55,12 @@ def main():
     os.environ['CA_PATH'] = ca_path
     openssl_cnf = os.path.join(ca_path, 'openssl.cnf')
 
+    country = os.path.expanduser(config.get('ca', 'country'))
+    state = os.path.expanduser(config.get('ca', 'state'))
+    locality = os.path.expanduser(config.get('ca', 'locality'))
+    organization = os.path.expanduser(config.get('ca', 'organization'))
+    organizational_unit = os.path.expanduser(config.get('ca', 'organizational_unit'))
+
     openssl = OpenSSL(opts.openssl_bin, openssl_cnf)
 
     if args[0] == 'init':
@@ -93,14 +99,18 @@ def main():
     if opts.password:
         password = opts.password
     if not login or not password:
-        p.error("No login details provided")
+        p.error("No iLO login details provided")
 
-    for hostname in args[1:]:
-        csr_path = os.path.join(ca_path, 'certs', hostname + '.csr')
-        crt_path = os.path.join(ca_path, 'certs', hostname + '.crt')
-        if os.path.exists(crt_path):
-            print("Certificate already signed, skipping")
-            continue
+    tries_count = 0
+    tries_count_max = 1
+    crt_requests = args[1:]
+    #for tries_count in range(tries_count_max):
+    pending_crt_requests = []
+    print("*"*80)
+    print("%s" % " ".join(crt_requests))
+    print("*"*80)
+
+    for hostname in crt_requests:
 
         ilo = hpilo.Ilo(hostname, login, password, opts.timeout, opts.port)
         ilo.debug = opts.debug
@@ -109,6 +119,19 @@ def main():
         elif opts.protocol == 'raw':
             ilo.protocol = hpilo.ILO_RAW
 
+        network_settings = ilo.get_network_settings()
+        dns_name = str(network_settings['dns_name'])
+        domain_name = str(network_settings['domain_name'])
+        ip_address = str(network_settings['ip_address'])
+        os.environ['NSSSLSERVERNAME'] = dns_name
+        os.environ['SUBJECTALTNAMERAW'] = 'DNS:%s,DNS:%s,IP:%s' % (ip_address, dns_name, ip_address)
+
+        csr_path = os.path.join(ca_path, 'certs', dns_name + '.csr')
+        crt_path = os.path.join(ca_path, 'certs', dns_name + '.crt')
+        if os.path.exists(crt_path):
+            print("Certificate already signed for %s, skipping" % hostname)
+            continue
+
         print("(1/5) Checking certificate config of %s" % hostname)
         fw_version = ilo.get_fw_version()
         if fw_version['management_processor'] == 'iLO2':
@@ -130,28 +153,37 @@ def main():
             if '.' not in cn:
                 ilo.cert_fqdn(use_fqdn=True)
                 cn = ilo.get_cert_subject_info()['csr_subject_common_name']
-            if cn != hostname:
-                print("Hostname (%s) and CN (%s) do not match, fixing" % (hostname, cn))
-                ilo.mod_network_settings(dns_name=hostname[:hostname.find('.')])
-                print("Waiting a minute to let the iLO reset itself")
-                time.sleep(60)
+            print("CN=%s, hostname=%s" % (cn, hostname))
+            #if cn != hostname:
+            #    print("Hostname (%s) and CN (%s) do not match, fixing" % (hostname, cn))
+            #    ilo.mod_network_settings(dns_name=hostname[:hostname.find('.')])
+            #    print("Waiting a minute to let the iLO reset itself")
+            #    time.sleep(60)
 
-        elif re.match(r'^iLO{3-9}$', fw_version['management_processor']):
-            cn = ilo.get_network_settings()['dns_name']
-            if cn != hostname[:hostname.find('.')]:
-                print("Hostname (%s) and CN (%s) do not match, fixing" % (hostname, cn))
-                ilo.mod_network_settings(dns_name=hostname[:hostname.find('.')])
-                print("Waiting a minute to let the iLO reset itself")
-                time.sleep(60)
+        elif re.match(r'^iLO[3-9]$', fw_version['management_processor']):
+            cn = dns_name + '.' + domain_name
+            print("*** CN=%s" % cn)
 
         else:
-            print("hpilo_ca cannot manage certificates for %s, only iLO2 and newer are supported" % (fw_version['management_processor'], hostname))
+            print("hpilo_ca cannot manage certificates for %s %s, only iLO2 and newer are supported" % (
+                fw_version['management_processor'], hostname))
             continue
 
+        # could iterate this upto 15 times and wait a minute between iterations, break when pending iterations is empty
         print("(2/5) Retrieving certificate signing request")
-        csr = ilo.certificate_signing_request()
+        try:
+            # this is a bit poor as I have hardcoded the certificate information, not sure that is works
+            if fw_version['management_processor'] == 'iLO2':
+                csr = ilo.certificate_signing_request()
+            else:
+                csr = ilo.certificate_signing_request(country=country, state=state, locality=locality, organization=organization, organizational_unit=organizational_unit, common_name=cn)
+        except hpilo.IloError as e:
+            print("%s: %s" % (hostname, str(e)))
+            if "The iLO subsystem is currently generating a Certificate Signing Request" in str(e):
+                pending_crt_requests.append(hostname)
+            continue
         if not csr:
-            print("Reveived an empty CSR")
+            print("Received an empty CSR")
             continue
         fd = open(csr_path, 'w')
         fd.write(csr)
@@ -170,6 +202,21 @@ def main():
         print("(5/5) Resetting iLO")
         ilo.reset_rib()
 
+
+    #if not crt_requests:
+    #    break
+
+    crt_requests = pending_crt_requests[:]
+    print("*"*80)
+    print("%s" % " ".join(crt_requests))
+    print("*"*80)
+
+    sys.exit(0)
+
+    print("Waiting one minute to let the iLO generate the signing request")
+    time.sleep(60)
+
+
 class OpenSSL(object):
     def __init__(self, openssl_bin, openssl_cnf):
         self.openssl_bin = openssl_bin
@@ -198,7 +245,7 @@ RANDFILE         = $dir/.rand
 x509_extensions  = usr_cert
 name_opt         = ca_default
 cert_opt         = ca_default
-default_days     = 1825 # 5 years
+default_days     = 1826 # 5 years
 default_crl_days = 30
 default_md       = sha1
 preserve         = no
@@ -240,13 +287,15 @@ commonName_default          = hpilo_ca
 basicConstraints       = CA:FALSE
 subjectKeyIdentifier   = hash
 authorityKeyIdentifier = keyid,issuer
-nsComment              = "Certificate generated by iLO CA"
+nsComment              = "Certificate by iLO CA"
+nsSslServerName        = $ENV::NSSSLSERVERNAME
+subjectAltName         = $ENV::SUBJECTALTNAMERAW
 
 [ca_cert]
 basicConstraints       = CA:true
 subjectKeyIdentifier   = hash
 authorityKeyIdentifier = keyid:always,issuer:always
-nsComment              = "Certificate generated by iLO CA"
+nsComment              = "Generated by iLO CA"
 """
 
 def print_progress(text):
```

**Comment by seveas** [comment]
Fixed in 5f6a076.
